src/Collections/Arrays/BinarySearch.dfy

```diff
@@ -6,17 +6,22 @@
 *******************************************************************************/
 
 include "../../Helpers.dfy"
+include "../../Wrappers.dfy"
+include "../../Relations.dfy"
 
 module Search {
   export
     provides BinarySearch, Wrappers, Helpers
-  
-  import Helpers, Comparison, Datatypes
-  import opened Wrappers = Helpers.Wrappers
+    provides Helpers, Wrappers 
 
-  method BinarySearch<T>(a: array<T>, key: T, compare: (T, T) -> bool) returns (r:Option<nat>)
-    requires forall i,j :: 0 <= i < j < a.Length ==> compare(a[i], a[j]) || a[i] == a[j]
-    requires Helpers.Trichotomous(compare)
+    import Helpers
+    import opened Relations
+    import opened Wrappers
+    import opened Comparison = Helpers.Comparison
+
+  method BinarySearch<T>(a: array<T>, key: T, compare: (T, T) -> Comparison.CompResult) returns (r:Option<nat>)
+    requires forall i,j :: 0 <= i < j < a.Length ==> compare(a[i], a[j]).LessThan? || a[i] == a[j]
```

Comment:
```suggestion
    requires forall i,j :: 0 <= i < j < a.Length ==> a[i] == a[j] || compare(a[i], a[j]).LessThan?
```
That reduces the number of comparisons if elements are mostly equal.

Reply:
Nah I wouldn't prematurely optimize like that - it also increases the number of comparisons if the elements are mostly not equal (which I would expect to be more common).

In fact this is ghost code, so why do we even care about extra operations? :)

Reply:
> In fact this is ghost code, so why do we even care about extra operations? :)

Right, that's a good point :-)